server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3418,34 +3431,48 @@ private <T> void executeShardSnapshotUpdate(
                     return;
                 }
 
-                final ShardSnapshotStatus updatedState;
+                final var newShardSnapshotStatusesBuilder = newShardSnapshotStatusesSupplier.get();
+                final var newShardSnapshotStatus = newShardSnapshotStatusesBuilder.get(shardSnapshotId);
+                if (newShardSnapshotStatus != null && newShardSnapshotStatus.state().completed()) {
```

Comment:
This should always be non-null - the builder starts with a copy of `existingShardSnapshotStatuses` and just overwrites the bits that get updated.